server/service/src/document/document_service.rs

```diff
@@ -0,0 +1,440 @@
+                if let Some(validator) = &validator {
+                    validate_json(&validator, &doc.data)
+                        .map_err(|errors| DocumentInsertError::InvalidDataSchema(errors))?;
+                }
+
+                match insert_document(con, store_id, doc) {
+                    Ok(doc) => Ok(doc),
+                    Err(err) => match err {
+                        DocumentInsertError::MergeRequired(ref merged_doc) => {
+                            // check that the merged document has a valid schema
+                            if let (Some(validator), Some(merged_doc)) = (&validator, merged_doc) {
+                                validate_json(&validator, &merged_doc.data)
+                                    .map_err(|_| DocumentInsertError::MergeRequired(None))?;
+                            }
+                            Err(err)
+                        }
+                        _ => Err(err),
+                    },
+                }
+            })
+            .map_err(|err| err.to_inner_error())?;
+        Ok(document)
+    }
+}
+
+pub struct DocumentService {}
+impl DocumentServiceTrait for DocumentService {}
+
+impl From<RepositoryError> for DocumentInsertError {
+    fn from(err: RepositoryError) -> Self {
+        match err {
+            RepositoryError::NotFound => DocumentInsertError::InvalidDocumentHistory,
+            _ => DocumentInsertError::DatabaseError(err),
+        }
+    }
+}
+
+fn two_way_document_merge(our: RawDocument, their: Document) -> RawDocument {
+    let our_data = our.data;
+    let their_data = their.data;
+
+    let solver = TakeLatestConflictSolver::new(our.timestamp.clone(), their.timestamp.clone());
+    let merged = two_way_merge(&our_data, &their_data, &solver);
+
+    let mut new_parents = our.parents;
+    new_parents.push(their.id);
+
+    RawDocument {
+        parents: new_parents,
+        timestamp: Utc::now(),
+        data: merged,
+
+        // keep exiting
+        name: our.name,
+        r#type: our.r#type,
+        author: our.author,
+        schema_id: our.schema_id,
+    }
+}
+
+fn three_way_document_merge(our: RawDocument, their: Document, base: Document) -> RawDocument {
+    let our_data = our.data;
+    let their_data = their.data;
+    let base_data = base.data;
+
+    let solver = TakeLatestConflictSolver::new(our.timestamp.clone(), their.timestamp.clone());
+    let merged = three_way_merge(&our_data, &their_data, &base_data, &solver);
+
+    let mut new_parents = our.parents;
+    new_parents.push(their.id);
+
+    RawDocument {
+        parents: new_parents,
+        timestamp: Utc::now(),
+        data: merged,
+
+        name: our.name,
+        author: our.author,
+        r#type: our.r#type,
+        schema_id: our.schema_id,
+    }
+}
+
+fn json_validator(
+    connection: &StorageConnection,
+    doc: &RawDocument,
+) -> Result<Option<JSONSchema>, DocumentInsertError> {
+    if let Some(schema_id) = &doc.schema_id {
```

Comment:
Another good use of returning schema_id  from match scope and early return from json_validator with Ok(None)